packages/styled-components/src/constructors/createGlobalStyle.js

```diff
@@ -12,22 +12,23 @@ import type { Interpolation } from '../types';
 
 type GlobalStyleComponentPropsType = Object;
 
-// place our cache into shared context so it'll persist between HMRs
-if (IS_BROWSER) {
-  window.scCGSHMRCache = {};
-}
-
 export default function createGlobalStyle(
   strings: Array<string>,
   ...interpolations: Array<Interpolation>
 ) {
   const rules = css(strings, ...interpolations);
   const styledComponentId = `sc-global-${hasher(JSON.stringify(rules))}`;
   const globalStyle = new GlobalStyle(rules, styledComponentId);
+  let count = 0;
 
   function GlobalStyleComponent(props: GlobalStyleComponentPropsType) {
     const styleSheet = useStyleSheet();
     const theme = useContext(ThemeContext);
+    const instanceRef = useRef(null);
+
+    if (instanceRef.current === null) instanceRef.current = String(++count);
```

Comment:
My only nit is that this doesn't have to be a `string`. It's concatenated to the ID (which is already a `string`) later anyway, so it should be fine to just keep the number.

This is also concurrent mode safe now, since if the render repeats without any effects, the number will just already be present.